### ISSUE TITLE
Implement extensible() support

### ISSUE THREAD

**Issue by matjaz** [firebase/bolt#30]
`extensible(prop) { return false; }` could be simply converted to `{"$other": {".validate": false}}`

`extensible(prop) { return prop == 'p1' || prop == 'p2'; }` is converted to `{"$other": {".validate":"$other == 'p1' || $other == 'p2'"}}`

There could also be simpler (CoffeeScript) syntax `extensible(prop) { return prop in ['p1', 'p2']; }`

**Comment by mckoss** [comment]
I'm working on this now.  Actually, extensible() is being removed, to be replaced with something like this:

    type Type {
      validate() { return $other.length < 10; }
      prop: String,
      $other: String
    }

**Comment by matjaz** [comment]
so no additional properties would be `$other: Null`
and mixed property types would be with `$other: String|number`
and more complex could be `validate() { return this.other1.isString() && (this.other2 == null || this.other2.isNumber()); }`
?

I would say with 'safe by default' there should be no additional properties by default.

**Comment by tomlarkworthy** [comment]
I like that line thinking @mckoss, pushing the validation into the Type system. however it's not clear as to whether the validate applies to fixed props or not. Also it seems at the wrong level. Shouldn't it be
```
validate() {
  return this[$other].length < 10;
}
```

However that runs into difficulty too because you might be setting several $other keys at the same time. You can't really mix the enclosing type's validate with the $other validator. You could push it all into the $other's type though, which is pretty neat I think. 

However, that means you can't perform validation on the key used. Type script solves typing dictionaries by typing the key too. So it would be more like:
```
type ParentType {
  validate() { /*parentType validation */}
  prop: String,
  $other: KeyType => ChildType 
}
```

**Comment by TristonianJones** [comment]
I like the direction as well @mckoss and @tomlarkworthy. The typing dictionary for additional properties seems like it would help considerably. With dynamic properties it seems possible to bind several:
```typescript
type DynamicType {
  prop: String,
  $otherObject: String => Object,
  $otherStringList: String => String[]
}
```

Dynamic types that contain many ignored properties seem like they might be harder to work with, but maybe I'm missing some aspect of the change.

**Comment by matjaz** [comment]
FB keys are always String.

Also we need to have an option to allow only specific key names. For example.

```typescript
type Email extends String {
  validate() { return this.indexOf('@') > 0; }
}
type Address {
  line1: String
  country: String
}

type Person {
  name: String
  $emails: Email
  $address: Address|Null
  $age: Number

  validate () { return $emails.startsWith('email') && $address.startsWith('address') && $age == 'age'; }
}
```

So that following object would be valid:
```json
{
  "name": "John Smith",
  "email1": "john@smith.com",
  "age": 42,
  "address1": {
    "line1": "Smith Str.",
    "country": "CA"
  }
}
```

**Comment by mckoss** [comment]
Doh.  Quite right, @tomlarkworthy - the validate is not scoped at the right level to include $other in the rule (even with this[$other] - which wasn't my intention for the sample - I was trying to limit the key length).

How about:

type Type {
  prop: String,
  $other: String {
    validate() { return $other.length < 10; }
  }
}

*BTW: this validation rule gets run (on the server) once for each property not matching the explicit property names of the Type.*

**Comment by tomlarkworthy** [comment]
I prefer the dictionary semantic which is matched with what a dynamic type is trying to achieve. (I also looked up how to do it properly in TS)

```
// create a type encoding custom key validation logic
type SmallString extends String {
   validate() {return this.length < 10;}
}

//now use that custom type as the key type in a dictionary
//typescript style embedded dictionaries
type Type {
  prop: String
  [key: SmallString]: Any
}

```

I like this because its a prexisting concept from Typescript, and it doesn't require a different code path for key validation vs. value validation i.e. it leverages the type validation to both situations. It does require a new dictionary concept though. (I am not convinced Typescript has the best syntax, I think I prefer => but that's not the point, the point is expressing key validation as just a parameter choice for construction of a parametric dictionary)

**Comment by matjaz** [comment]
So how would one define type Person where two dynamic properties are related?
```typescript
type Person {
  name: String
  [email: SmallString]: Email
  [address: SmallString]: Address
  validate() { return this.email.endsWith('google.com') && this.address.zip == 94043; }
}
```

**Comment by mckoss** [comment]
My plan is to implement  collection-like syntax as a valid property type.  This would not give access to writing validation rules based on the $key directly in the type statement (you'd have to write a parallel validation rule in a path for that).

To rewrite matjaz's example:

```
type Email extends String {
  validate() { return this.indexOf('@') > 0; }
}

type Address {
  line1: String,
  country: String
}

type Person {
  name: String,
  emails: Emails[],
  addresses: Addresses[],
  age: Number | Null,
}

path /people is Person[] {}
```

**Comment by matjaz** [comment]
I think following safe by default, types should be by default freezed (no additional properties), but still allowing type to be easily extensible.

For example for game board we have (X&Y) coordinates. from 1,1 to 10,10. I should be able to easily check if coordinate is already taken. So I need a hash.

```
type BoardMove {
  userId: UserId
  ts: Timestamp
}

path /gamesBoard/$gameId/moves is BoardMove[] {} // no need for that
path /gamesBoard/$gameId/moves/$coordinate is BoardMove {
  validate() { return $coordinate.test(/^\d,\d$/); }
}
```

I think of a two different scenarios for dynamic properties (ones not explicitly defined).
To validate if dynamic property key is acceptable. (above example)
To validate if dynamic property value is acceptable. (for example in above example if timestamp is less than 10seconds before previous.)

**Comment by tomlarkworthy** [comment]
@matjaz 
So obviously that example is not great as you shouldn't mix dynamic properties, but here is the dictionary approach for comparison.

- note I assume we are using the email as the key in emails, but there is no clear value, so I will jsut put a boolean in it (validated or not?).

```
type Email extends String {
  validate() { return this.indexOf('@') > 0; }
}
type Address {
  line1: String,
  country: String
}

type Person {
  name: String,
  {Email => Boolean} |    // the type of this, is the union of two dynamic types (+ fixed fields)
  {String => Address}
  age: Number | Null,
}

path /people/$name is Person {}
OR
path /people is {String => Person}
```

I kinda think using arrays is the wrong thing as we don't use arrays in Firebase. We use dictionaries.

**Comment by mckoss** [comment]
Firebase supports generic type Object<String, Any>.  Developers use this as the way to support internal collections (when just defined in a property but not in their own path statement).

But `emails: Object<String, Email>` seemed rather verbose for a common paradigm (though I do support this type expression in my latest PR)....perhaps we should prefer `Map<String, Email>`?

So I propose to make `Email[]` an alias for that type expression. Other aliases options: `List<Email>`, `Collection<Email>`, `Object<Email>`.

You'll note that Firebase does allow for Javascript Arrays to be written to a Object<String, Map> (and in fact, Javascript actually implements the internal Array type as just that!).

I looked at [Swift type expressions](https://developer.apple.com/library/ios/documentation/Swift/Conceptual/Swift_Programming_Language/CollectionTypes.html) - they use [Email] for Array<Email> and [String: Email] for Dictionary<String, Email> - I'm not a fan.

**Comment by matjaz** [comment]
Here's another idea...

```typescript
type BoardCoordinate extends String { // in Firebase keys are strings
  validate() { return this.test(/^\d,\d$/); }
}
```

```typescript
path /gamesBoard/$gameId/moves is BoardCoordinate => BoardMove {
  validate() { return this.numChildren() < 42 && this.hasChild('0,0'); }
}
```

It's not enough to validate hash values but ALSO keys. And as in above example also some additional control for how many children hash can have or that some key is mandatory etc...

**Comment by mckoss** [comment]
FYI - Firebase has no numChildren function available - (presumably because it could be a very expensive operation in the general case).

I'm not a fan of the Key => Value notation (as we may use this for "arrow functions") - I assume what is meant here is Object<BoardCoordinate, BoardMove> (using generics notation).

**Comment by matjaz** [comment]
there is https://www.firebase.com/docs/web/api/datasnapshot/numchildren.html

what about
```
type BoardMoves {
  '0,0': BoardMove
  $coordinate: Map<BoardCoordinate, BoardMove>
  validate() { return this.numChildren() < 42 }
}
```

**Comment by mckoss** [comment]
I stand corrected!

**Comment by mckoss** [comment]
We DO have numChildren - implemented in the clients - it copies all data in child node to execute it.  But no such method in the [Security and Rules API](https://www.firebase.com/docs/security/api/).

**Comment by mckoss** [comment]
Question is - is `$prop: Type` necessary?  Your example could be:

```
path /gamesBoard/$gameId/moves is Map<BoardCoordinate, BoardMove>;
```

**Comment by matjaz** [comment]
It's not the same.
With my example you have two additional conditions:
1. property '0,0' is mandatory
2. numChildren() (or any other) additional validation (e.g. sum of fields)

**Comment by mckoss** [comment]
numChildren just isn't supported on the server Rules API (to which we are translating), I'm afraid.

What do you think of this for the required child case?

```
path /gamesBoard/$gameId/moves is Map<BoardCoordinate, BoardMove> {
  validate() { return this['0,0'] != null; }
}
```

**Comment by matjaz** [comment]
yeah doable. But I would really like to see this in type. For code reuse.

**Comment by mckoss** [comment]
Except we don't have $prop: Type.

**Comment by mckoss** [comment]
You should be able to reuse a type like this:

```
type Moves extends Map<BoardCoordinate, BoardMove> {
  validate() { return this['0,0'] != null; }
}

path /gamesBoard/$gameId/moves is Moves;
```
But this is Not Yet Implemented ... but I THINK this will work.

**Comment by matjaz** [comment]
That's also great.

**Comment by tomlarkworthy** [comment]
Yeah the parametric Map works as well. I think that looks quite understandable for most people. +1

**Comment by mckoss** [comment]
@matjaz - See https://github.com/firebase/bolt/pull/57 - generic types support (most of) what we discussed here.

**Comment by mckoss** [comment]
Resolved by f0b108fcdae66ffdee82f6e01b77589d59ddc572
